topsites.py

```diff
@@ -0,0 +1,180 @@
+
+
+class TopSites:
+    '''Used to query top site info from alexa API to Site object'''
+    def __init__(self, access_key, secret_key):
+        self.access_key = access_key
+        self.secret_key = secret_key
+
+    def query(self, begin, begin_priority):
+        '''Query alexa with given range, save to Site'''
+        query_string = self.build_query(begin, 100)
+        toSign = "GET\n" + ATS_SERVICE_HOST + "\n/\n" + query_string
+        signature = self.gen_sign(toSign, self.secret_key)
+        uri = "{base}{query}&{signature}".format(
+            base=ATS_AWS_BASE_URL,
+            query=query_string,
+            signature=urlencode({"Signature": signature}, "UTF-8"))
+
+        response = requests.get(uri)
+
+        if response.status_code == requests.codes.ok:
+            # See http://docs.aws.amazon.com/AlexaTopSites/latest/index.html?QUERY_QueryRequests.html  # nopep8
+            dom = parseString(response.content)
+            sites = dom.getElementsByTagName("aws:Site")
+            for site in sites:
+                url = self.get_text(
+                    site.getElementsByTagName("aws:DataUrl")[0].childNodes)
+                rank = int(self.get_text(
+                    site.getElementsByTagName("aws:Rank")[0].childNodes))
+
+                priority = begin_priority
+                if rank > 100 and rank <= 1000:
+                    priority = priority + 1
+                elif rank > 1000 and rank <= 10000:
+                    priority = priority + 2
+
+                if url in topsites:
+                    site_row = topsites[url]
+                    # Update priority is new is lower
+                    if site_row.priority > priority:
+                        site_row.priority = priority
+                        site_row.country_code = self.country_code
+                        site_row.ranking = rank
+                else:
+                    site_row = Site(url, priority, self.country_code, rank)
+                    topsites[url] = site_row
+                    session.add(site_row)
+
+            session.commit()
+
+    def build_query(self, start, count):
+        '''Build query string for request with start ranking and count'''
+        now = time.time()
+        mlsec = repr(now).split('.')[1][:3]
+        nowtime = datetime.datetime.utcnow()
+        timestamp = nowtime.strftime(ATS_DATEFORMAT).format(mlsec)
+
+        query_params = {
+            "Action": ATS_ACTION_NAME,
+            "ResponseGroup": ATS_RESPONSE_GROUP_NAME,
+            "AWSAccessKeyId": self.access_key,
+            "Timestamp": timestamp,
+            "CountryCode": self.country_code,
+            "Count": count,
+            "Start": start,
+            "SignatureVersion": "2",
+            "SignatureMethod": ATS_HASH_ALGORITHM
+        }
+
+        # alexa top site only accept request with ordered query parameters
+        return urlencode(collections.OrderedDict(sorted(query_params.items())))
+
```

Comment:
Here I'm trying to understand.
```python
>>> query_params = {
...     "Action": ATS_ACTION_NAME,
...     "ResponseGroup": ATS_RESPONSE_GROUP_NAME,
...     "AWSAccessKeyId": "blah",
...     "Timestamp": "2017-06-14T05:50:09.480Z",
...     "CountryCode": "US",
...     "Count": 100,
...     "Start": 234,
...     "SignatureVersion": "2",
...     "SignatureMethod": ATS_HASH_ALGORITHM
... }
>>> 
>>> # This will give you a list of tuples.
>>> query_params.items()
[('Count', 100), ('Start', 234), ('SignatureVersion', '2'), ('AWSAccessKeyId', 'blah'), ('CountryCode', 'US'), ('Action', 'TopSites'), ('Timestamp', '2017-06-14T05:50:09.480Z'), ('SignatureMethod', 'HmacSHA256'), ('ResponseGroup', 'Country')]
>>> # This is sorting the list of tuple.
>>> sorted(query_params.items())
[('AWSAccessKeyId', 'blah'), ('Action', 'TopSites'), ('Count', 100), ('CountryCode', 'US'), ('ResponseGroup', 'Country'), ('SignatureMethod', 'HmacSHA256'), ('SignatureVersion', '2'), ('Start', 234), ('Timestamp', '2017-06-14T05:50:09.480Z')]
>>> import collections
>>> # This converts it to an ordered dictionary
>>> collections.OrderedDict(sorted(query_params.items()))
OrderedDict([('AWSAccessKeyId', 'blah'), ('Action', 'TopSites'), ('Count', 100), ('CountryCode', 'US'), ('ResponseGroup', 'Country'), ('SignatureMethod', 'HmacSHA256'), ('SignatureVersion', '2'), ('Start', 234), ('Timestamp', '2017-06-14T05:50:09.480Z')])
>>> from urllib import urlencode
>>> # This encodes it as a parameter URL.
>>> urlencode(collections.OrderedDict(sorted(query_params.items())))
'AWSAccessKeyId=blah&Action=TopSites&Count=100&CountryCode=US&ResponseGroup=Country&SignatureMethod=HmacSHA256&SignatureVersion=2&Start=234&Timestamp=2017-06-14T05%3A50%3A09.480Z'
>>> # But this would work too.
>>> urlencode(sorted(query_params.items()))
'AWSAccessKeyId=blah&Action=TopSites&Count=100&CountryCode=US&ResponseGroup=Country&SignatureMethod=HmacSHA256&SignatureVersion=2&Start=234&Timestamp=2017-06-14T05%3A50%3A09.480Z'
```

But why not make it even simpler. You are not using the dictionary for its benefit, aka deferencing with keys a value. We could use simply a list of tuples. List in python are ordered objects.

```python
>>> access_key = 'foo'
>>> count = 100
>>> country_code = 'US'
>>> start = '333'
>>> start = 333
>>> timestamp = '2017-06-14T05:50:09.480Z'
>>> query_params = [
...     ("Action", ATS_ACTION_NAME),
...     ("AWSAccessKeyId", access_key),
...     ("Count", count),
...     ("CountryCode", country_code),
...     ("ResponseGroup", ATS_RESPONSE_GROUP_NAME),
...     ("SignatureMethod", ATS_HASH_ALGORITHM),
...     ("SignatureVersion", "2"),
...     ("Start", start),
...     ("Timestamp", timestamp)]
>>> urlencode(query_params)
'Action=TopSites&AWSAccessKeyId=foo&Count=100&CountryCode=US&ResponseGroup=Country&SignatureMethod=HmacSHA256&SignatureVersion=2&Start=333&Timestamp=2017-06-14T05%3A50%3A09.480Z'
```

The full function could become.
```python
    def build_query(self, start, count):
        ATS_DATEFORMAT = '%Y-%m-%dT%H:%M:%S.%f'
        now_time = datetime.datetime.utcnow()
        timestamp = '{}Z'.format(nowtime.strftime('ATS_DATEFORMAT')[:-3])
        # Alexa top site only accept request with ordered query parameters
        # Keep the order!
        query_params = [
            ("Action", ATS_ACTION_NAME),
            ("AWSAccessKeyId", access_key),
            ("Count", count),
            ("CountryCode", country_code),
            ("ResponseGroup", ATS_RESPONSE_GROUP_NAME),
            ("SignatureMethod", ATS_HASH_ALGORITHM),
            ("SignatureVersion", "2"),
            ("Start", start),
            ("Timestamp", timestamp)]
        return urlencode(query_params)
```